### PR TITLE
Create newtype for platform triple in bootstrap.

### PR DESCRIPTION
**Pull request by frewsxcv**
As I was reading through the bootstrap code, I noticed myself getting
lost at times trying to find out what the value of each `&str` is. I
also noticed there's quite a few platform-checking conditionals checking
the contents of these strings. In an attempt to clarify and cleanup
these cases, I created a 'newtype' `Triple` that represents a platform
triple and moved the platform-checking logic to methods on the new
struct.

### CODE REVIEW

File: src/bootstrap/triple.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+use std::fmt;
+use std::path;
+
+/// Platform represented by a hyphen-delimited string. Does not
+/// necessarily consist of three items. e.g.:
+///
+/// * `i686-apple-darwin`
+/// * `powerpc-unknown-linux-gnu`
+#[derive(Clone, Debug, Default, Eq, Hash, PartialEq, RustcDecodable)]
+pub struct Triple(pub String);
```
Comment (frewsxcv) [inline src/bootstrap/triple.rs]: In the future, maybe this could be an `enum` that has all the platforms we support?

Comment (TimNN) [inline src/bootstrap/triple.rs]: I don't think that would work due to the flexible / json target specifications.

**Comment by frewsxcv** [comment]
Rebased.

**Comment by alexcrichton** [comment]
Thanks for the PR! I've been somewhat hesitant to do this in the past though as I've not been super certain about what the benefit is and being wary of over-abstracting. This is already something that all build scripts have to deal with (e.g. they don't have a shared `Triple` type).

Could you elaborate though on some of the confusion you saw? I wonder if perhaps a `pub struct Triple(str)` could help where we just `Deref` to `str` to have a separate type

**Comment by frewsxcv** [comment]
When I see a `&str` as a parameter, if it's not documented, I have to do some work to find out what the underlying content is and how it's formatted. If I see a `Triple`, I know exactly what sort of data I'm working with. To me, the newtype pattern acts like a compile-time comment; I find it especially helpful on loose types like `&[u8]` or `&str`.

Anyways, I don't feel strongly about this, mostly just throwing around patterns I personally find helpful.

**Comment by alexcrichton** [comment]
Hm yeah it's true that `&str` isn't descriptive, but it's also what all build scripts are using anyway I think? (e.g. the "source of truth" is just a string). In that sense I hoped that names like `target` and `host` would convey "target triple", but perhaps the docs could be bolstered as well?

**Comment by frewsxcv** [comment]
> it's also what all build scripts are using anyway

Just to clarify, which build scripts are you referring to?

**Comment by alexcrichton** [comment]
Oh just build scripts in general for Cargo crates, which the build system is somewhat intended to follow after

**Comment by frewsxcv** [comment]
I'm not seeing anything about a target triple in the [build script documentation](http://doc.crates.io/build-script.html), so I'm still not entirely sure what you're referring to here. Regardless, it doesn't sound like there's much enthusiasm around this change, so I'm leaning towards closing this. I'm curious to know your thoughts about the ['newtype' pattern](https://doc.rust-lang.org/book/structs.html#tuple-structs) since to me, this seemed like a good use of it.

**Comment by alexcrichton** [comment]
Oh sorry I just meant env vars like `$TARGET` and `$HOST`.

**Comment by alexcrichton** [comment]
I'm personally not a huge fan of newtypes as extra types in the limit just ends up forcing everyone to juggle a huge number of types in their heads, but when applied well they can certainly work nicely!

**Comment by frewsxcv** [comment]
Sounds good, thanks for the input!

**Pull request closed without merging**